tasks/apply-mapping/apply-mapping.yaml

```diff
@@ -62,44 +62,71 @@ spec:
             exit 0
         fi
 
-        # Expected arguments are [tags, timestamp, release_timestamp, git sha, 7 character git sha, source sha]
-        translate_tags () {
+        # Expected arguments are: [variable, substitute_map, labels_map]
+        substitute() {
+            variable=$1
+            substitute_map=$2
+            labels_map=$3
+
+            result=""
+            if [[ "$variable" == labels.* ]]; then
+                label="${variable#labels.}"
+                result="$(jq -r --arg labelval "$label" '.[$labelval] // ""' <<< "${labels_map}")"
+                # result="$(jq -r --arg label "$label" '.[$label] // ""' <<< "{}")"
+            else
+                result="$(jq -r --arg variable "$variable" '.[$variable] // ""' <<< "${substitute_map}")"
+            fi
+            echo "$result"
+        }
+
+        # Expected arguments are [tags, substitute_map, labels_map]
         # The tags argument is a json array
-            if [[ $1 == '' ]] ; then
+        translate_tags () {
+            tags=$1
+            substitute_map=$2
+            labels_map=$3
+            if [ "$tags" = '' ] ; then
                 echo ''
                 return
             fi
 
-            SUPPORTED_VARIABLES='[
-                {"{{timestamp}}": "'$2'"},
-                {"{{ timestamp }}": "'$2'"},
-                {"{{release_timestamp}}": "'$3'"},
-                {"{{ release_timestamp }}": "'$3'"},
-                {"{{git_sha}}": "'$4'"},
-                {"{{ git_sha }}": "'$4'"},
-                {"{{git_short_sha}}": "'$5'"},
-                {"{{ git_short_sha }}": "'$5'"},
-                {"{{digest_sha}}": "'$6'"},
-                {"{{ digest_sha }}": "'$6'"}
-            ]'
-            tags=$1
+            translated_tags='[]'
+            NUM_TAGS="$(jq 'length' <<< "${tags}")"
+            for ((i = 0; i < NUM_TAGS; i++)); do
+                tag="$(jq -r --argjson i "$i" '.[$i]' <<< "${tags}")"
 
-            NUM_VARIABLES=$(jq 'length' <<< "${SUPPORTED_VARIABLES}")
-            for ((i = 0; i < NUM_VARIABLES; i++)); do
-                variable=$(jq -c --argjson i "$i" '.[$i]' <<< "${SUPPORTED_VARIABLES}")
-                KEY=$(jq -r 'to_entries[] | .key' <<< "$variable")
-                VALUE=$(jq -r 'to_entries[] | .value' <<< "$variable")
-                # If a tag is going to be replaced by a null value, error out
-                if [[ "$VALUE" == "null" ]] && \
-                  [[ $(jq --arg key "$KEY" '. | map(select(test($key))) | length' <<< "$tags") -gt 0 ]] ; then
-                    echo "Requested substitution for key: $KEY but calculated value for it was null"
-                    echo "Failing..."
+                while [[ $tag =~ \{\{\ *([[:alnum:]_\.]+)\ *\}\} ]]; do
```

Comment:
I have faith this is correct, but can you say what this is doing? Replying here would help, adding a comment would be better

Reply:
I added a comment:
```
# Repeatedly translate {{}} references until none are left
```
Let me know if it's still not clear.